wordpress-mobile/WordPress-FluxC-Android
# Add API endpoint support for adding custom packages/activating predefined packages.

### Description:
As WooCommerce Shipping & Tax now [has API support for creating/activating packages with POST](https://github.com/Automattic/woocommerce-services/pull/2318), this PR updates FluxC to be able to make use of the endpoint.

The immediate need of it is for WooCommerce Android to support package creation / activation in the app. Related issues: 
- [Support creating custom packages](https://github.com/woocommerce/woocommerce-android/issues/3993)
- [Support activating predefined packages](https://github.com/woocommerce/woocommerce-android/issues/3992)

In general there are two types of packages:
1. **Custom package**: Here users create and enter their own package specification.
2. **Predefined packages**: This is supplied by core and is a list of packages provided by the supported carriers. Here the endpoint enables activation of such packages.

This PR adds support for both creating a custom package, and activating a predefined package, as well as add buttons to test them in the example app.

This PR also includes support for creating multiple and mixed package types at once, although tests for these cases do not currently exist in the example app, only within unit tests &mdash; I have only tested them manually.

<hr />

### Testing Instruction:
1. Run example app
2. Tap **Woo**, then tap **Shipping Labels**
2. Tap **Select Site**, and pick a site that already has **WooCommerce Shipping & Tax** plugin enabled and activated
3. Scroll down and find the two buttons: **Create A Custom Package** and **Activate a Random Predefined Package**
4. Tap the **Create A Custom Package** button and complete the creation steps. If all goes well, it will say `Custom package created: true` in the log area
6. Check in WooCommerce Shipping settings area in wp-admin (`wp-admin/admin.php?page=wc-settings&tab=shipping&section=woocommerce-services-settings`) and confirm that the newly created custom package is listed under the **Packaging** section
7. Go back to example app and tap the  **Activate a Random Predefined Package** button.
8. The button will try to activate a predefined package randomly. The log will say `Activating [package_id] from [carrier_name]`. <br> **8.a:** If it goes through correctly, the log will then say `Predefined package activated: true`.<br>  **8.b:** There's a possibility this does not go through if the selected package is already activated. The log will say `GENERIC_ERROR: At least one of the new predefined packages has the same name as existing packages.` If this happens, retry the button again until it picks something that goes through.
9. Check in WooCommerce Shipping settings area in wp-admin again, and confirm that the newly activated package is listed under Packaging (compare the name with the `[package_id]` listed in the log).

Aside from the testing steps above, run unit tests on `WCShippingLabelStoreTest.kt` and ensure everything passes.
